lua/rainbow/internal.lua

```diff
@@ -34,54 +34,44 @@ local function color_no(mynode, len)
         end
 end
 
-local callbackfn = function(bufnr, parser)
-        -- no need to do anything when pum is open
-        if vim.fn.pumvisible() == 1 then
-                return
-        end
+local callbackfn = function(bufnr, _, changes, tree, lang)
+        --vim.schedule_wrap(function()
+                -- no need to do anything when pum is open
+                if vim.fn.pumvisible() == 1 or not lang then
+                        return
+                end
 
-        --clear highlights or code commented out later has highlights too
-        vim.api.nvim_buf_clear_namespace(bufnr, nsid, 0, -1)
-        parser:parse()
-        parser:for_each_tree(function(tree, lang_tree)
-                local root_node = tree:root()
+                for _, change in ipairs(changes) do
 
-                local lang = lang_tree:lang()
-                local query = queries.get_query(lang, "parens")
-                if query ~= nil then
-                        for _, node, _ in query:iter_captures(root_node, bufnr) do
-                                -- set colour for this nesting level
-                                local color_no_ = color_no(node, #colors)
-                                local _, startCol, endRow, endCol = node:range() -- range of the capture, zero-indexed
-                                vim.highlight.range(
-                                        bufnr,
-                                        nsid,
-                                        ("rainbowcol" .. color_no_),
-                                        { endRow, startCol },
-                                        { endRow, endCol - 1 },
-                                        "blockwise",
-                                        true
-                                )
+                        ----clear highlights or code commented out later has highlights too
+                        vim.api.nvim_buf_clear_namespace(bufnr, nsid, change[1], change[3])
```

Comment:
Maybe only clear range if there is at least one capture match in the change?

Reply:
Yeah but why is clearing it unnecessarily bad?

Reply:
I was thinking about the situation when you are typing within a word and only adding letters. Then you probably don't want to check all parens. But probably this won't even trigger `on_treechanged` because it only adds bytes and does not change the structure of the tree